presto-main/src/main/java/com/facebook/presto/cost/DisjointRangeDomainHistogram.java

```diff
@@ -0,0 +1,359 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.spi.statistics.ConnectorHistogram;
+import com.facebook.presto.spi.statistics.Estimate;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.BoundType;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeSet;
+import com.google.common.collect.TreeRangeSet;
+
+import java.util.Collection;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.cost.HistogramCalculator.calculateFilterFactor;
+import static com.facebook.presto.util.MoreMath.max;
+import static com.facebook.presto.util.MoreMath.min;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.lang.Double.NEGATIVE_INFINITY;
+import static java.lang.Double.POSITIVE_INFINITY;
+import static java.lang.Double.isFinite;
+import static java.util.Objects.hash;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class represents a set of disjoint ranges that span an input domain.
+ * Each range is used to represent filters over the domain of an original
+ * "source" histogram.
+ * <br>
+ * For example, assume a source histogram represents a uniform distribution
+ * over the range [0, 100]. Next, assume a query with multiple filters such as
+ * <code>x < 10 OR x > 75</code>. This translates to two disjoint ranges over
+ * the histogram of [0, 10) and (75, 100], representing roughly 35% of the
+ * values in the original dataset. Using the example above, a cumulative
+ * probability for value 5 represents 5% of the original dataset, but 20% (1/5)
+ * of the range of constrained dataset. Similarly, all values in [10, 75] should
+ * compute their cumulative probability as 40% (2/5).
```

Comment:
Are the numbers in this example inaccurate?

Reply:
These numbers are accurate

Reply:
I'm a little confused that value 5 got a cumulative probability 20% (1/5) on dataset [0, 10) and (75, 100], which spans 35. Shouldn't it got 1/7 on a dataset which spans 35, or got 1/5 on a dataset which spans 25? Any misunderstandings please let me know, thanks.

Reply:
Ah you are correct. It should have been 85. Updated